### ISSUE TITLE
Copy propagation cleanup

### ISSUE THREAD

**Issue by jstolarek** [jstolarek/ghc#24]
There's a lot of clutter in my copy-propagation pass. Mostly debugging output. After that debugging is removed transfer and rewrite functions can be eta-reduced.

There's also a question of what to do with the preprocessing pass before copy propagation.

And another question whether I have to erase facts about the stack during unsafe foreign calls.

 1) we don't need to track information about stack facts, because we are only using stack to 
       a) get function params from the stack; 
       b) pass params on the stack before exiting from a function.
     If we did copy-prop after stack layout then we would also use stack to save live parameters across calls, but we could not propagate these. So we don't need to track stack facts. Unless we missed something?

  2) Sinking pass does most of the job we wanted copy-propagation to do: it inlines expressions that are used only once, inlines usages of local registers and eliminates dead assignments. There are a couple of things that sinking pass does not yet do. We decided with Simon that I will improve sinking so that it is more robust and then copy propagation shouldn't be necessary.

**Comment by jstolarek** [comment]
performance, tests, code cleanup, document on the wiki, dead code in loops, LLVM impact, throw out loopification, general performance.

**Comment by jstolarek** [comment]
This is no longer relevant, as my copy propagation pass will not be merged into HEAD.
